sqlx-postgres/src/types/cube.rs

```diff
@@ -0,0 +1,393 @@
+use crate::decode::Decode;
+use crate::encode::{Encode, IsNull};
+use crate::types::Type;
+use crate::{PgArgumentBuffer, PgTypeInfo, PgValueFormat, PgValueRef, Postgres};
+use sqlx_core::Error;
+
+const BYTE_WIDTH: usize = 8;
+const CUBE_TYPE_ZERO_VOLUME: usize = 128;
+const CUBE_TYPE_DEFAULT: usize = 0;
+const CUBE_DIMENSION_ONE: usize = 1;
+const DIMENSIONALITY_POSITION: usize = 3;
+const START_INDEX: usize = 4;
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum PgCube {
+    Point(f64),
+    ZeroVolume(Vec<f64>),
+    OneDimensionInterval(f64, f64),
+    MultiDimension(Vec<Vec<f64>>),
+}
+
+impl Type<Postgres> for PgCube {
+    fn type_info() -> PgTypeInfo {
+        PgTypeInfo::with_name("cube")
+    }
+}
+
+impl<'r> Decode<'r, Postgres> for PgCube {
+    fn decode(value: PgValueRef<'r>) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
+        match value.format() {
+            PgValueFormat::Text => Ok(PgCube::try_from(value.as_str()?)?),
+            PgValueFormat::Binary => Ok(PgCube::try_from(value.as_bytes()?)?),
+        }
+    }
+}
+
+impl<'q> Encode<'q, Postgres> for PgCube {
+    fn produces(&self) -> Option<PgTypeInfo> {
+        Some(PgTypeInfo::with_name("cube"))
+    }
+
+    fn encode_by_ref(&self, buf: &mut PgArgumentBuffer) -> IsNull {
+        buf.extend_from_slice(&self.serialize());
+
+        IsNull::No
+    }
+}
+
+impl TryFrom<&str> for PgCube {
+    type Error = Error;
+
+    fn try_from(input: &str) -> Result<Self, Self::Error> {
+        let content = &input.get(1..input.len() - 1).ok_or(Error::Decode(
```

Comment:
What characters is this meant to be trimming here? It's surprising to do so unconditionally.

Reply:
This is for the open and closed parens at the start and end of the string representation.
E.g. '(1, 2)' or '((1,2), (3,4))'

Reply:
According to the docs, the string format may contain a bare float as a one-dimensional point: https://www.postgresql.org/docs/current/cube.html

Postgres may always return a single point wrapped with parens, but since this is exposed as a `TryFrom<&str>` impl, users may try to pass a bare number as a single point.

Reply:
Should I just be implementing this as a function then so that we're not exposing any traits? Also to align with the u8 buf deserialising.

Reply:
I could see the `FromStr` impl being really convenient for constructing a `PgCube`, so I'm okay with keeping it public. But if we want to keep it public, we should accept any format that Postgres would consider to be valid.